src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -482,88 +484,79 @@ private void ProcessAssignmentOperation(IAssignmentOperation assignmentOperation
             {
-                sinkKinds = null;
-
-                if (method.ContainingType == null || !taintedArguments.Any())
-                {
-                    return false;
-                }
-
-                Lazy<HashSet<SinkKind>> lazySinkKinds = new Lazy<HashSet<SinkKind>>(() => new HashSet<SinkKind>());
+                IEnumerable<IArgumentOperation> taintedArguments = null;
                 foreach (SinkInfo sinkInfo in this.DataFlowAnalysisContext.SinkInfos.GetInfosForType(method.ContainingType))
                 {
-                    if (lazySinkKinds.IsValueCreated && lazySinkKinds.Value.IsSupersetOf(sinkInfo.SinkKinds))
+                    taintedArguments ??= GetTaintedArguments(visitedArguments);
+                    if (taintedArguments == null)
                     {
-                        continue;
+                        return;
                     }
 
-                    if (method.MethodKind == MethodKind.Constructor
-                        && sinkInfo.IsAnyStringParameterInConstructorASink
-                        && taintedArguments.Any(a => a.Parameter.Type.SpecialType == SpecialType.System_String))
-                    {
-                        lazySinkKinds.Value.UnionWith(sinkInfo.SinkKinds);
-                    }
-                    else if (sinkInfo.SinkMethodParameters.TryGetValue(method.MetadataName, out ImmutableHashSet<string> sinkParameters)
-                        && taintedArguments.Any(a => sinkParameters.Contains(a.Parameter.MetadataName)))
+                    foreach (IArgumentOperation taintedArgument in taintedArguments)
                     {
-                        lazySinkKinds.Value.UnionWith(sinkInfo.SinkKinds);
+                        if ((method.MethodKind == MethodKind.Constructor
+                            && sinkInfo.IsAnyStringParameterInConstructorASink
+                            && taintedArguments.Any(a => a.Parameter.Type.SpecialType == SpecialType.System_String))
+                            || (sinkInfo.SinkMethodParameters.TryGetValue(method.MetadataName, out ImmutableHashSet<string> sinkParameters)
+                            && taintedArguments.Any(a => sinkParameters.Contains(a.Parameter.MetadataName)))
+                            || (originalOperation is IInvocationOperation invocationOperation
+                            && invocationOperation.Instance != null
+                            && this.GetCachedAbstractValue(invocationOperation.Instance).Kind == TaintedDataAbstractValueKind.Tainted
+                            && sinkInfo.SinkMethodParametersWithTaintedInstance.TryGetValue(method.MetadataName, out ImmutableHashSet<string> sinkMethodParametersWithTaintedInstance)
+                            && taintedArguments.Any(a => sinkMethodParametersWithTaintedInstance.Contains(a.Parameter.MetadataName))))
+                        {
```

Comment:
Would be good to add indentation to help make clear the structure of the logic here (e.g. which && and || are associated with what)